src/v/io/scheduler.cc

```diff
@@ -0,0 +1,162 @@
+         * semaphore. later a unit is deposited into the semaphore and the first
+         * wakes up and opens its underlying io_queue. at this point the running
+         * queue monitor could be suspended, but we'd like to be able for all
+         * queues to make progress. to handle this the active queue monitor will
+         * evict one or more other queues before suspending itself. however,
+         * because an io queue needs to wake up to be closed there is a delay
+         * before the eviction is visible. this delay is why we track waiters_
+         * separately rather than using the nofiles_ semaphore state.
+         */
+        while (waiters_ > 0 && !lru_.empty()) {
+            waiters_--;
+            auto& queue = lru_.front();
+            lru_.pop_front();
+            queue.sem_.signal();
+        }
+
+        /*
+         * when stopping the queue clean it up fully including removing it from
+         * the cache and closing the underlying io-queue.
+         */
+        if (queue->stop_) {
+            if (queue->io_queue_.opened()) {
+                if (queue->cache_hook_.is_linked()) {
+                    queue->cache_hook_.unlink();
+                }
+                co_await queue->io_queue_.close();
+                queue->units_.return_all();
+            }
+            co_return;
+        }
+
+        co_await queue->sem_.wait(std::max<size_t>(queue->sem_.current(), 1));
+
+        /*
+         * in the common case the queue is already open so there is nothing else
+         * to do and the monitor can loop around to suspend itself. however, if
+         * the queue has also been removed from the lru cache list then this is
+         * a signal that the opened queue should be closed and nofiles_ units
+         * released.
+         */
+        if (queue->io_queue_.opened()) {
+            if (!queue->cache_hook_.is_linked()) {
+                co_await queue->io_queue_.close();
+                queue->units_.return_all();
+            }
+            continue;
+        }
+
+        /*
+         * open the queue. if we can't get nofiles units then we'll try to
+         * evict another queue and wait for units to become available.
+         */
+        auto units = seastar::try_get_units(nofiles_, 1);
+        if (!units.has_value()) {
+            if (lru_.empty()) {
+                /*
+                 * this case is possible, if for example, `nofiles_` was
+                 * initialized with 0 units.
+                 */
+                waiters_++;
+            } else {
+                /*
+                 * signal another queue to be closed. once closed this queue
+                 * will acquire a nofiles_ unit and be allowed to open.
+                 */
+                auto& queue = lru_.front();
+                lru_.pop_front();
+                queue.sem_.signal();
+            }
+            units = co_await seastar::get_units(nofiles_, 1);
```

Comment:
do I understand correctly that the invariant is that if the queue is in the `lru_` list it should have units so if we are waking up the queue it's guaranteed that we will eventually acquire units on this line?

Reply:
yes. but i think as you point out, a queue may still hold on to the units even if it isn't on the `lru` list. this is a temporary state, and not being on the list is a signal to the queue monitor that when it runs it should release its units _after_ stopping the io_queue and closing the file.